internal/vm/remotevm/builder.go

```diff
@@ -0,0 +1,126 @@
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/vm"
+	"github.com/Microsoft/hcsshim/internal/vmservice"
+	"github.com/containerd/ttrpc"
+	ptypes "github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var _ = (vm.UVMBuilder)(&utilityVMBuilder{})
+
+type utilityVMBuilder struct {
+	id      string
+	guestOS vm.GuestOS
+	job     *jobobject.JobObject
+	config  *vmservice.VMConfig
+	client  vmservice.VMService
+}
+
+func NewUVMBuilder(ctx context.Context, id, owner, binPath, addr string, guestOS vm.GuestOS) (vm.UVMBuilder, error) {
+	var job *jobobject.JobObject
+	if binPath != "" {
+		log.G(ctx).WithFields(logrus.Fields{
+			"binary":  binPath,
+			"address": addr,
+		}).Debug("starting remotevm server process")
+
+		opts := &jobobject.Options{
+			Name: id,
+		}
+		job, err := jobobject.Create(ctx, opts)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create job object for remotevm process")
+		}
+
+		cmd := exec.Command(binPath, "--ttrpc", addr)
+		p, err := cmd.StdoutPipe()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create stdout pipe")
+		}
+
+		if err := cmd.Start(); err != nil {
+			return nil, errors.Wrap(err, "failed to start remotevm server process")
+		}
+
+		if err := job.Assign(uint32(cmd.Process.Pid)); err != nil {
+			return nil, errors.Wrap(err, "failed to assign remotevm process to job")
+		}
+
+		if err := job.SetTerminateOnLastHandleClose(); err != nil {
+			return nil, errors.Wrap(err, "failed to set terminate on last handle closed for remotevm job object")
+		}
+
+		f, err := os.Open(os.DevNull)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to open nul file")
+		}
+
+		// Wait for stdout to close. This is our signal that the server is successfully up and running.
+		_, _ = io.Copy(f, p)
+	}
+
+	conn, err := net.Dial("unix", addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to dial remotevm address %q", addr)
+	}
+
+	c := ttrpc.NewClient(conn, ttrpc.WithOnClose(func() { conn.Close() }))
+	vmClient := vmservice.NewVMClient(c)
+
+	return &utilityVMBuilder{
+		id:      id,
+		guestOS: guestOS,
+		config: &vmservice.VMConfig{
+			MemoryConfig:    &vmservice.MemoryConfig{},
+			DevicesConfig:   &vmservice.DevicesConfig{},
+			ProcessorConfig: &vmservice.ProcessorConfig{},
+			SerialConfig:    &vmservice.SerialConfig{},
+			ExtraData:       make(map[string]string),
+		},
+		job:    job,
+		client: vmClient,
+	}, nil
+}
+
+func (uvmb *utilityVMBuilder) Create(ctx context.Context) (vm.UVM, error) {
+	// Grab what capabilities the virtstack supports up front. Do a quick check right after to
+	// see if the guest OS is supported.
+	capabilities, err := uvmb.client.CapabilitiesVM(ctx, &ptypes.Empty{})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get virtstack capabilities from vmservice")
+	}
+
+	supported := false
+	for _, supportedOS := range capabilities.SupportedGuestOs {
```

Comment:
Oh this was more so if the vmm didn't support booting a "insert OS here". e.g. if Linux guests were supported but not Windows and we were getting asked to launch a Windows VM.

Reply:
It seems weird to ask the VMM which OSes it can boot, though. The shim knows more about the OS that we're trying to boot than the VMM, so it's in a better position to determine whether it can boot given the VMM's capabilities. E.g., a VMM might have the right devices to boot Windows in some configurations, but it still won't be able to boot a Windows UVM if it doesn't have vSMB.

So I think it would be better either to skip this check, or to check something like: if windows && (!vsmbsupported || !uefisupported) { return nil, err }

Reply:
Ok thanks for this, that makes more sense :). On it

Reply:
Should we just remove that section off of the CapabilitiesVMResponse? We don't have anyway to tell if X thing is supported on initial creation of a VM from the response so it needs some changes regardless

Reply:
(not in this PR)

Reply:
I'm going to remove this check for now, and let's work out what CapabilitiesVM should look like and reinvestigate later.